### ISSUE TITLE
Tendermint BFT Delta

### ISSUE THREAD

**Issue by aneequesafdar** [clearmatics/autonity#347]
# Problem

The current implementation of Tendermint doesn't completely adhere to the [latest gossip on BFT consensus](https://arxiv.org/abs/1807.04938v3) paper and therefore the implementation is not byzantine fault-tolerant, we have to make the following changes to reach bft (may need to be split up in separate issues):

- Change the implementation of line 55 of algorithm 1 to treat different types of future messages from the same validator as 1 towards the total count of the set.
- Update committed seals with signatures of Precommits as described in issue #202.
- [`TimeoutEvent`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L21) needs to be removed because when a timer expires in algorithm 1 of the paper relevant timeout function is called which usually send a message or starts a new round.
- The upon conditions on lines 28, 36 and 49 in algorithm 1 needs to be checked in both message handlers. This requires that at any point in time during the height the validator needs to have access to all of the round states for the current height.

## Moving to higher round (Line 55)

In the current implementation of line 55, a node will move to a higher round if it receives f+1 messages of round higher ([`futureRoundChange`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/handler.go#L319)). This meant that a node will count a Proposal, Prevote and Precommit as 3 separate messages towards the total count of f+1 messages, this is extremely problematic because a byzantine validators can cause honest nodes to change round prematurely. The intuition behind f+1 messages is that at least one honest validator would have sent a correct message, however, this is predicated on receiving distinct messages from validators depending on the context, when we allow for Proposals, Prevotes and Precommits to count as separate messages from the same validator towards the total count of f+1 we are reducing the threshold of the future messages required to change the round.

Consider the example where n=4 and there is 1 byzantine validator while the rest are honest, therefore, f=1 and f+1=2. All of the validators start a new height and are in round 0, the byzantine validator decides to send a pevote and Precommit of round 10 to one of the honest validator and not others, this will cause the honest validator to jump to round 10 while the other honest validators will remain in round 0, and now the byzantine validator goes offline leaving different validator in different rounds, none of them can perform a round change without receiving 2f+1 messages to start the timeout, thus no progress can be made. In the best-case scenario, if through gossip honest validator receive the messages from the byzantine and also move to round 10 (all honest validator are in round 10), the byzantine validator can send Prevotes and Precommits for another higher round before honest validator have a chance to send Proposals, Prevotes and Precommits, thus byzantine validator can constantly keep changing rounds and cause honest validator to get stuck in livelock, where rounds are progressing but the state of the blockchain doesn't change.

The solution to this problem is counting Proposals, Prevotes and Precommits from a validator for a future round as one towards the total count of f+1. For example, if a validator receives a Prevote and Precommit from another validator for a higher round, then it should be counted as one message instead of two towards the total count of messages needed to change to that round. Line 55 in algorithm 1 of the paper describes the set as `<*, h_p, round, *, *>` and the property of the set is that there are no duplicates since `<PROPOSAL, h_p, round, *, *>`, `<PREVOTE, h_p, round, *, *>` and `<PRECOMMIT, h_p, round, *, *>` all fit the pattern of `<*, h_p, round, *, *>` while the only differentiating factor of the message is `<>`, which serves as the signature of the sender and therefore we count Proposal, Prevote and Precommits as duplicates in the context of moving to future rounds.

In practice, this means that every time we handle a future round message and add it to the relevant round state we need to update the `futureRoundChange` according to the rule above. `futureRoundChange` is a map of round numbers to the number of future messages we have received during the current height and once we hit the f+1 threshold we will change to that future round.

## Update Committed Seals (#202)

The current implementation of the committed seal is `SignECDSA(Keccak256(CONCAT(precommit.ProposalBlockHash, msgPrecommit)), PrivateKey)` ([`PrepareCommittedSeal()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/precommit.go#L62)), this is a major bug because Proposal block hash can remain the same across rounds and thus committed seals will also remain same across rounds. This can lead to a byzantine proposer collecting committed seals from different validators across different rounds and committing a block using committed seals from different rounds violating line 49 of algorithm 1.

The committed seal needs to be a signature of the validator of the Precommit message. The message contains all the relevant information about the round, height, Precommit and address of the sender, therefore, a committed seal of an older round cannot be used as a committed seal of a higher round. This also requires the Precommit messages or votes (depending on the implementation) needs to be added to the extra data field, because, to verify the signature we need the content the validator is signing. In the current implementation the signature is over the Proposal hash and a constant, therefore, it is very easy to recreate the content signature is signing from the rest of the block header and there is no need to add Precommit messages to the extra data field. This can be observed in the [`verifyCommittedSeals()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/backend/engine.go#L224).

Therefore, when creating the list of committed seals in [`commit()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/core.go#L241) we need to add the relevant message or vote to the extra data field. When we receive a block we need to update the [`verifyCommittedSeals()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/backend/engine.go#L224) such that we check the commit seal with the corresponding Precommit message in the extra data field.

## Remove Timeout Event

The [`TimeoutEvent`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L21) is sent when a timer expires and it contains the relevant information about which handler to run: [`handleTimeoutPropose()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L160), [`handleTimeoutPrevote()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L168) or [`handleTimeoutPrecommit()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L176). These time out handlers will send a `Prevote(nil), Precommit(nil) or start a new round`, however, according to lines 21, 35 and 48 of algorithm 1 we need to call the relevant handler as soon as the timer expires.

In the current implementation, we unnecessarily wait longer after the timer has expired to send the relevant message since we first send and handle the time out event. The timeout handlers are simply going to send another message event anyways, so there is no need to send timeout event in the first place rather we should simply send the relevant message on timer expiry. Also since the [`eventLoop`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/handler.go#L164) will randomly choose between `MessageEvent` or `TimeoutEvent` (if both cases are ready at the same time) we can delay in executing timeout logic, this will deviate from lines 21, 35 and 48 of algorithm 1.

Therefore, we need to remove `TimeoutEvent` and the timeout handlers, replace the logic in [`onTimeoutPropose()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L126), [`onTimeoutPrevote()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L137) and [`onTimeoutPrecommit()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/timeout.go#L148) as follows (not exact implementation changes, just for illustration purposes):

```go
func (c *core) onTimeoutPropose(ctx context.Context, r int64, h int64) {
 if h == c.currentRoundState.Height().Int64() && r == c.currentRoundState.Round().Int64() && c.currentRoundState.Step() == propose {
 c.logTimeoutEvent("TimeoutEvent(Propose): Received", "Propose", msg)
 c.sendPrevote(ctx, true)
 c.setStep(prevote)
 }
}

func (c *core) onTimeoutPrevote(ctx context.Context, r int64, h int64) {
 if h == c.currentRoundState.Height().Int64() && r == c.currentRoundState.Round().Int64() && c.currentRoundState.Step() == prevote {
 c.logTimeoutEvent("TimeoutEvent(Prevote): Received", "Prevote", msg)
 c.sendPrecommit(ctx, true)
 c.setStep(precommit)
 }
}

func (c *core) onTimeoutPrecommit(ctx context.Context, r int64, h int64) {
 if h == c.currentRoundState.Height().Int64() && r == c.currentRoundState.Round().Int64() {
 c.logTimeoutEvent("TimeoutEvent(Precommit): Received", "Precommit", msg)
 c.startRound(ctx, new(big.Int).Add(c.currentRoundState.Round(), common.Big1))
 }
}

```

We must make sure that this feature is implemented properly so that honest validator do not send both `Prevote(B)/Precommit(B)` and `Prevote(nil)/Precommit(nil)`, this should be ensured by step changes and locks on the round states.

## Access to all round states of the current height and common message conditions

The Tendermint algorithm described in algorithm 1 of the latest gossip on BFT consensus paper is based on assumptions and abstraction used to describe distributed algorithms. This makes it difficult to express or stay close to the conventions of the algorithm during implementation. The description of the algorithm assumes an infinitely large set of collective messages on which all of these upon conditions are based on. There is no mention of how these upon condition are checked for and it assumes a mechanism for adding to the infinitely large message set.

In practice, the algorithm has to be implemented from the perspective of a single validator (i.e. `p` becomes fixed). Thus, a state needs to be derived from variables mentioned in the algorithm description and a mechanism is required to store the messages as well as which messages are required at a given state. In algorithm 1 of the paper, the height is always in terms of the validator since all mentions of height are referred to as `h_p`. Therefore, we can safely assume that at any given moment during the execution of the Tendermint the validator can ignore messages from older and future heights, also since messages can be lost (as we allow for crash and recovery), ignoring these messages can be seen as lost messages if they are required these can be requested through round state synchronisation. However, a validator cannot discard or ignore old and future round messages since line 28 of the algorithm requires access to previous round Prevotes because valid round (`vr`) is less than the validator's round while line 49 predicates messages on any round to fulfil the condition.

In the current implementation, [`roundState`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/roundstate.go#L40) object is used to keep track of all the variables defined at initialisation of the algorithm 1 as well as the messages related to that round. There are multiple round states to manage old round states ([`currentHeightOldRoundsStates`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/core.go#L145)) but it is not correctly implemented to handle all types of messages. Future round messages are stored in the [`backlogs`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/core.go#L124). Backlogs is a map of validators to a priority queue which contains unverified messages, this makes it very difficult to have access to future round messages needed to either move to a future round or to check for the condition on line 49.

The solution to this problem is to deprecate backlog and process all of the messages we receive as long as they are of the same height as the validator. We also need to deprecate [`currentHeightOldRoundsStates`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/core.go#L145) and instead update `roundState` to keep track of all the rounds' messages for the current height (probably by using a map of the round number to message set). Thus we will only have one `roundState` for the whole height and will no longer require us to create new `roundState`s for each round. `lockedRound`, `validRound`, `lockedValue` and `validValue` should be moved from [`core`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/core.go#L102) to `roundState`. The new `roundState should look like the following (not exact implementation changes, just for illustration purposes):

```go
type roundState struct {
 round *big.Int
 height *big.Int

 step Step

 lockedRound *big.Int
 validRound *big.Int
 lockedValue *types.Block
 validValue *types.Block

 roundMessages map[int]roundMessageSet
 mu sync.RWMutex
}
```

As mentioned above, the description of the algorithm doesn't describe how different messages are received nor how the upon event conditions are checked. The Proposal, Prevote and Precommit messages are the only ones responsible for changing the state of a validator, therefore it makes sense to check for the upon conditions after receiving any of these messages. The simplest strategy would be to check for all of the upon conditions after receiving any message, however, a better strategy is to only check for upon conditions which are related to that message type. For example, line 22 of the algorithm only refers to a Proposal message so this upon condition should not be check when a Prevote or a Precommit message is received. Unfortunately, this strategy is not correctly implemented in Autonity, for each message type a handler is defined which is meant to check for these upon conditions, however, common upon conditions of different messages are not checked in each handler, namely, lines 28, 36 and 49.

- Proposal Upon Conditions

  - Line 22: `upon <PROPOSAL, h_p , round_p , v, −1> from proposer(h_p , round_p ) while step_p = propose`
  - Line 28: `upon <PROPOSAL, h_p , round_p , v, vr> from proposer(h_p , round_p ) AND 2f + 1 <PREVOTE, h_p , vr, id(v)> while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p )`
  - Line 36: `upon <PROPOSAL, h_p , round_p , v, ∗> from proposer(h_p , round_p ) AND 2f + 1 <PREVOTE, h_p , round_p , id(v)> while valid(v) ∧ step_p ≥ prevote for the first time`
  - Line 49: `upon <PROPOSAL, h_p , r, v, ∗> from proposer(h_p , r) AND 2f + 1 <PRECOMMIT, h_p , r, id(v)> while decision_p[h_p] = nil`

- Prevote Upon Conditions

  - Line 28: `upon <PROPOSAL, h_p , round_p , v, vr> from proposer(h_p , round_p ) AND 2f + 1 <PREVOTE, h_p , vr, id(v)> while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p )`
  - Line 34: `upon 2f + 1 <PREVOTE, h_p , round_p , ∗> while step_p = prevote for the first time`
  - Line 36: `upon <PROPOSAL, h_p , round_p , v, ∗> from proposer(h_p , round_p ) AND 2f + 1 <PREVOTE, h_p , round_p , id(v)> while valid(v) ∧ step_p ≥ prevote for the first time`
  - Line 44: `upon 2f + 1 <PREVOTE, h_p , round_p , nil> while step_p = prevote`

- Precommit Upon Conditions

  - Line 47: `upon 2f + 1 <PRECOMMIT, h_p , round_p , ∗> for the first time`
  - Line 49: `upon <PROPOSAL, h_p , r, v, ∗> from proposer(h_p , r) AND 2f + 1 <PRECOMMIT, h_p , r, id(v)> while decision_p[h_p] = nil`

The common condition between Proposal and Prevote messages are lines 28 and 36, while the common condition between Proposal and Precommit messages is line 49. Therefore, lines 28 and 36 must be checked in both [`handleProposal()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/propose.go#L54) and [`handlePrevote()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/prevote.go#L61), while line 49 must be checked in both [`handleProposal()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/propose.go#L54) and [`handlePrecommit()`](https://github.com/clearmatics/autonity/blob/de53f1f346ef2f8f29d55529bf83a0aaf51fc809/consensus/tendermint/core/precommit.go#L72).

Line 28 needs to be checked in Proposal and Prevote handler because we may have a Proposal where the valid round in more than -1 but the Prevote messages of the valid round (which will be an old round by definition) are yet to arrive (we know 2f+1 should arrive eventually because the proposer must have seen 2f+1 Prevotes messages for the block to re-propose it) so as soon as the valid round Prevotes messages arrive we need to check for this condition in the Prevote handler to ensure we unlock (assuming the validator is still in the propose step and Propose timeout has not expired).

Line 36 needs to checked in Proposal and Prevote handlers because it is possible to receive a valid Proposal after the propose step (i.e no valid Proposal was received before the propose timeout expiry) and after 2f+1 Prevotes, therefore, when the valid Proposal arrives we need to check for this condition in Propose handler.

Line 49 needs to be checked in Proposal and Precommit handler because it is possible to receive 2f+1 Precommits before receiving a valid Proposal, therefore, when a valid Proposal arrives we need to check in the Proposal handler for this condition. Also, this line should always have the highest priority while checking for the upon conditions since we want to decide as soon as possible and commit a block.

Line 28 and 36 require the validator to be in a specific step to execute corresponding code when these upon condition are true. Therefore, it is possible to have the first part of the condition true (the part before `while` statement) but the validator is not in the correct step when the messages arrive. For example, 2f+1 Prevote messages (let's assume no other Prevotes will arrive) for the valid Proposal arrive before the valid Proposal in the propose step and once the valid Proposal arrives the step will be changed to prevote. Since no other Prevote message will arrive the Prevote handler will not be called and there will be no check for line 36. This is wrong because we have enough message and are in the correct step to check and execute line 36. Hence, it is not sufficient only rely on Proposal, Prevote and Precommit handlers to check for all the upon conditions.

Upon conditions which do not predicate on a certain step, message handlers are sufficient to check for these conditions. Upon conditions which do predicate on certain steps, those conditions need to be checked when the step change is made to ensure the true conditions are executed as soon as possible. In our current implementation, no such check is made when steps are changed, there is a strong assumption on when the messages will be received. We assume that a Proposal, Prevotes and Precommits messages will arrive in their respective step since there are no checks for what happens when a message arrives outside of their step. This goes against the algorithm 1 of the paper because it is based on executing the algorithm when the upon conditions become true and it's not a sequential execution.

The following describes which upon conditions should be checked when a validator moves to a certain step:

- Propose Step Upon Conditions

  - Line 22: `upon <PROPOSAL, h_p , round_p , v, −1> from proposer(h_p , round_p ) while step_p = propose`
  - Line 28: `upon <PROPOSAL, h_p , round_p , v, vr> from proposer(h_p , round_p ) AND 2f + 1 <PREVOTE, h_p , vr, id(v)> while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p )`

- Prevote Step Upon Conditions

  - Line 34: `upon 2f + 1 <PREVOTE, h_p , round_p , ∗> while step_p = prevote for the first time`
  - Line 36: `upon <PROPOSAL, h_p , round_p , v, ∗> from proposer(h_p , round_p ) AND 2f + 1 <PREVOTE, h_p , round_p , id(v)> while valid(v) ∧ step_p ≥ prevote for the first time`
  - Line 44: `upon 2f + 1 <PREVOTE, h_p , round_p , nil> while step_p = prevote`

- Precommit Step Upon Conditions

  - Line 36: `upon <PROPOSAL, h_p , round_p , v, ∗> from proposer(h_p , round_p ) AND 2f + 1 <PREVOTE, h_p , round_p , id(v)> while valid(v) ∧ step_p ≥ prevote for the first time`

The algorithm doesn't describe an order in which the conditions should be executed when multiple of them are ready, rather it is chosen randomly. Having a certain order for checking and executing the condition (which is necessary during implementation as we don't want to introduce randomness while choosing which condition to execute) can be seen as an event with a low probability. The safety and liveness properties will not be affected by a certain order of execution. Most of the upon conditions are mutually exclusive, usually, there is only one upon condition true at a given time. Line 49 is the only condition that can be true at any time with another upon condition since it doesn't predicate on a specific round nor a step, however, intuitively, it should always be given the highest preference since we want to commit a block as soon as possible. The other non mutually exclusive conditions are line 34 and 44, 34 and 36, both of the cases are dealing with timeouts and it doesn't matter which condition is run first because line 34 will start a timeout and soon after line 36 or 44 would we run, making the timeout redundant (assuming both conditions were true at the same time).

In summary, the problem we are trying to solve how is to ensure the correct upon conditions are run as soon as the conditions become true since the algorithm doesn't describe how to handle messages nor does it describe the mechanism when to check for these conditions, it becomes the responsibility of the developer to make these decisions while making sure the algorithm is correctly implementing the distributed systems abstractions.

**Comment by aesmonty** [comment]
> line 36 and 44 are mutually exclusive
> line 34 and 44, and line 36 and 44 are not mutually exclusive

There is a problem here @aneequesafdar. Maybe a typo. They are mutually exclusive.

**Comment by aneequesafdar** [comment]
yeah you are right it should have been `34 and 44, 34 and 36 are not mutually exclusive`. And yes 36 and 44 are mutually exclusive

**Comment by JekaMas** [comment]
I think we still need a list of BFT tests that should be ok after this issue. Could we discuss this list?
